### DISCUSSION TITLE
Writing tests with state$

### DISCUSSION THREAD

**Discussion by evertbouw** [redux-observable/redux-observable#793] [Ideas]
Hey there 👋 

I'd like to gather examples for testing epics with the new state stream. This way we can help people with the upgrade to 1.0 and make it easy for newcomers to get started. Post what you've got or what you are struggling with!

Here's what I came up with so far
https://gist.github.com/evertbouw/d4c6e588c6ffd46febfdee6ac9d33878

**Comment by evertbouw** [comment]
Made a version that uses the TestScheduler method by @jayphelps https://github.com/ReactiveX/rxjs/pull/3596. Inspired by @clivend in https://github.com/redux-observable/redux-observable/issues/144

https://gist.github.com/evertbouw/94d4a9fdce93b34c31614a673acab73f

I've only written a handful of tests with this so I don't know yet if/how/when it breaks.

**Comment by evertbouw** [comment]
I've edited the above example so it is now possible to use a marble diagram for your states. This is very useful to check that `withLatestFrom` actually grabs the state at the time you expect. Switch the `delay` with `withLatestFrom` and whoops!

**Comment by huy-nguyen** [comment]
I'm pretty execited about `state$` so I wrote a [blog post](https://www.huynguyen.io/marble-diagrams-redux-observable-2018-05/?utm_source=github&utm_medium=ref&utm_campaign=githubissue20180527) to show beginners how to do marble testing with it. Any feedback is welcome!

**Comment by frankwallis** [comment]
@evertbouw @huy-nguyen many thanks for your work and the above examples, I am currrently moving some existing marble tests over to use state$ and rxjs6 and I do have some questions/feedback:

1) observables created via marble syntax cannot easily be made into StateObservable

If I want to represent the state changing over time I can use the TestScheduler to create a mock Observable like so:

```
const state$ = m.hot('s--tuv', stateMarbleValues);
```
But this will be `Observable<State>` not `StateObservable<State>` and so if I pass it into my epic, I cannot access `$state.value` as it will not exist. `StateObservable` constructor requires a `Subject` and so there are type errors if you try and construct one with a mock Observable. Looking at the code it seems that the `StateObservable` constructor could just take an `Observable` or even `Subscribable` instead?

2) When creating `state$` and `action$` using the TestScheduler it is not obvious whether to use `hot` or `cold` - I actually think they should be hot?

This becomes important when trying to model state changes over time with late subscriptions (e.g. when using `retry` or `repeat` or even `switchMap`). In @huy-nguyen's examples the state never changes over time and so the issue does not arise, but I suspect late subscribers would be receiving the entire stream of states rather than just the states emitted after they subscribe.

3) Related to (2) and not really about testing - it might be helpful to provide a convenience operator which just emits the single latest state and then completes. Most of the examples I have seen use `withLatestFrom` to combine the latest value from `state$` with another observable, but sometimes (again when using `retry` or `repeat`) you just want to get the latest state. Because of (1) I am avoiding $state.value in my epics and so instead of using `of($state.value)` I am using constructions like this:
```
   of({}, scheduler).pipe(withLatestFrom(state$, (_, value) => value))
```
or possibly 
```
   state$.pipe(shareReplay(1, Infinity, scheduler), take(1));
```
I haven't quite figured it out.

**Comment by evertbouw** [comment]
(Updated my gist to reduce boilerplate. Didn't realise it would be possible to use the same values object for all marbles. Thanks @huy-nguyen !)

> 1. observables created via marble syntax cannot easily be made into StateObservable

In my gist you can see that I'm using `const state$ = new StateObservable(hot(states, values));`. This also works with the rxjs-marbles library:
```js
const pingEpic = (action$, state$) => action$.pipe(
  ofType("ping"),
  map(() => ({ type: "pong", payload: state$.value })),
);

it('should work with state$.value', marbles(m => {
    const values = {
      s: "Hello world!",
      a: { type: "ping" },
      z: { type: "pong", payload: "Hello world!" },
    };
    const state$ = new StateObservable(m.hot('s', values));
    const action$ = m.hot('a', values);
    const expected$ = m.hot('z', values);

    const actual$ = pingEpic(action$, state$);

    m.expect(actual$).toBeObservable(expected$);
}));
```

> 2. When creating state$ and action$ using the TestScheduler it is not obvious whether to use hot or cold - I actually think they should be hot?

From what I can tell the only difference is that in hot observables you can use the `^` character to indicate the point of subscription. I've not needed this.

> Related to (2) and not really about testing - it might be helpful to provide a convenience operator which just emits the single latest state and then completes.

Yes there are a couple of operators I use a lot so combining them as part of the library makes a lot of sense. Again I want it to work for almost everyone and as easy as possible 😄 If you've found a pattern feel free to make a PR!

**Comment by frankwallis** [comment]
@evertbouw thanks for the response, I hadn't noticed the second gist.. That is exactly what I was looking for. I also see now that HotObservable is a Subject so the type error I was seeing must have been due to something else. I'm making some progress now :)

**Comment by julioxavierr** [comment]
Thank you @evertbouw, docs should be much more clear about this.

**Comment by julioxavierr** [comment]
Can we add this to the official docs? I would be happy to help
@evertbouw @jayphelps @benlesh

**Comment by rgbkrk** [comment]
Yes please add this to the official docs!
